rosbag2_performance/rosbag2_performance_benchmarking/src/benchmark_publishers.cpp

```diff
@@ -20,79 +20,118 @@
 #include "rosbag2_performance_benchmarking/byte_producer.hpp"
 #include "rosbag2_performance_benchmarking/config_utils.hpp"
 #include "rosbag2_performance_benchmarking/publisher_group_config.hpp"
+#include "rosbag2_performance_benchmarking/thread_pool.hpp"
+
+#include "msg_utils/message_producer_factory.hpp"
 
 #include "rclcpp/executors/single_threaded_executor.hpp"
 #include "rclcpp/node.hpp"
 #include "rclcpp/qos.hpp"
-#include "std_msgs/msg/byte_multi_array.hpp"
-
 
 class BenchmarkPublishers : public rclcpp::Node
 {
 public:
   explicit BenchmarkPublishers(const std::string & name)
   : rclcpp::Node(name)
   {
-    configurations_ = config_utils::publisher_groups_from_node_parameters(*this);
-    if (configurations_.empty()) {
-      RCLCPP_ERROR(get_logger(), "No publishers/producers found in node parameters");
-      return;
-    }
-
-    create_benchmark_publishers_and_producers();
+    create_benchmark_producers();
+    wait_for_subscriptions();
   }
 
   void run()
   {
-    std::vector<std::thread> producer_threads;
+    thread_pool_.start(std::thread::hardware_concurrency());
```

Comment:
This is a hidden variable within the benchmark apparatus which is not easily inspected or controlled for.